daemon/logger/loggerutils/rotatefilewriter.go

```diff
@@ -80,24 +85,86 @@ func (w *RotateFileWriter) checkCapacityAndRotate() error {
 	return nil
 }
 
-func rotate(name string, maxFiles int) error {
+func rotate(name string, maxFiles int, compress string) error {
 	if maxFiles < 2 {
 		return nil
 	}
-	for i := maxFiles - 1; i > 1; i-- {
-		toPath := name + "." + strconv.Itoa(i)
-		fromPath := name + "." + strconv.Itoa(i-1)
+
+	extension := ""
+	var compressionAlg archive.Compression
+	if compress != "" {
+		switch compress {
+		case "gzip":
+			compressionAlg = archive.Gzip
+			extension = ".gz"
+		case "bzip2":
+			compressionAlg = archive.Bzip2
+			extension = ".bz"
+		case "xz":
+			compressionAlg = archive.Xz
+			extension = ".xz"
+		default:
+			return fmt.Errorf("unknown compression algorithm %q for json-file", compress)
+		}
+	}
+
+	for i := maxFiles - 1; i > 2; i-- {
+		toPath := name + "." + strconv.Itoa(i) + extension
+		fromPath := name + "." + strconv.Itoa(i-1) + extension
 		if err := os.Rename(fromPath, toPath); err != nil && !os.IsNotExist(err) {
 			return err
 		}
 	}
 
+	if _, err := os.Stat(name + ".1"); err == nil && maxFiles > 2 {
+		if err := os.Rename(name+".1", name+".2"); err != nil {
+			return err
+		}
+
+		if err := compressFile(name+".2", compressionAlg, extension); err != nil {
+			return err
+		}
+	}
+
+	// The "[name].1" that jast renamed from "[name]" is not compressed
+	// in order to prevent the log tracking tool from losing some historical
+	// log data when a new log file is created.
 	if err := os.Rename(name, name+".1"); err != nil && !os.IsNotExist(err) {
 		return err
 	}
 	return nil
 }
 
+func compressFile(fileName string, compression archive.Compression, extension string) (err error) {
+	outFile, err := os.OpenFile(fileName+extension, os.O_CREATE|os.O_RDWR, 0640)
+	defer func() {
+		outFile.Close()
+		if err != nil {
+			os.Remove(fileName + extension)
+		}
+	}()
+
+	if err != nil {
+		return err
+	}
+
+	compressWriter, err := archive.CompressStream(outFile, compression)
+	defer compressWriter.Close()
+
+	fileData, err := ioutil.ReadFile(fileName)
```

Comment:
I don't think we want to read this in here. This should stream to the compressed writer.